### PR TITLE
Use stored procedures for metrics calculation

### PR DESCRIPTION
**Pull request by nscuro**
### Description

Updating metrics currently is a time-intensive task when dealing with large portfolios.

One of the reasons why it takes longer than necessary is the large number of unnecessary network round-trips between API server and database. Data is fetched, aggregated, and written back to the database. Metrics updates do not require any information from outside the database, so they can and should be performed directly on the database server itself.

One of the ideas shared in https://github.com/DependencyTrack/dependency-track/issues/1210#issuecomment-991692019 involved usage of (materialized) views, as project- and portfolio metrics really are just aggregates. While useful, views will only ever represent the "current" state. But DT needs to track metrics over time, so a refresh of metrics always involves either:

* Updating of an existing data point (only the `LAST_OCCURRENCE` column), or
* Creating of a new data point

A compromise between views and the current Java-based implementation are [stored procedures](https://www.postgresql.org/docs/current/xproc.html). They too are executed directly on the database, but are also able to perform `INSERT`s and `UPDATE`s.

> **Warning**  
> Stored procedures in SQL syntax are not supported in H2. This means that metrics updates will no longer work when using the embedded H2 database. I believe this breaking change is OK, since Hyades will require an external database anyway.

This PR replaces the previous Java-based implementation of metrics updates with stored procedures. Currently, only PostgreSQL is covered. In the future, we'll have to support MSSQL and MySQL as well.

The implementation behaves just like the previous one. However, in a future change we may decide to *not* update all project metrics when updating the overall portfolio metrics. Project metrics are updated on multiple occasions anyway, there's not really a point in forcing a refresh every X hours.

### Addressed Issue

N/A

### Additional Details

Each stored procedure runs in its own transaction. This is implicit behavior.

Due to this, the `UPDATE_PROJECT_METRICS` procedure does **not** call `UPDATE_COMPONENT_METRICS` for every component, and `UPDATE_PORTFOLIO_METRICS` does **not** call `UPDATE_PROJECT_METRICS` for every project. Transactions would simply grow too large, and block for too long.

It is thus still necessary to call each procedure separately from Java.

### Checklist

- [x] I have read and understand the [contributing guidelines](../CONTRIBUTING.md#pull-requests)
- ~This PR fixes a defect, and I have provided tests to verify that the fix is effective~
- [x] This PR implements an enhancement, and I have provided tests to verify that it works as intended
- ~This PR introduces changes to the database model, and I have added corresponding [update logic](https://github.com/DependencyTrack/dependency-track/tree/master/src/main/java/org/dependencytrack/upgrade)~
- ~This PR introduces new or alters existing behavior, and I have updated the [documentation](https://github.com/DependencyTrack/dependency-track/tree/master/docs/_docs) accordingly~

### CODE REVIEW

File: src/main/java/org/dependencytrack/tasks/metrics/PortfolioMetricsUpdateTask.java [hunk 1]
```diff
@@ -52,37 +50,53 @@ public class PortfolioMetricsUpdateTask implements Subscriber {
...
+        }
+    }
+
+    private static void refreshProjectMetrics() throws Exception {
         try (final var qm = new QueryManager().withL2CacheDisabled()) {
             final PersistenceManager pm = qm.getPersistenceManager();
 
             LOGGER.debug("Fetching first " + BATCH_SIZE + " projects");
-            List<Project> activeProjects = fetchNextActiveProjectsPage(pm, null);
+            List<ProjectProjection> activeProjects = fetchNextActiveProjectsPage(pm, null);
```
[the remark this answers is not in the document]
Comment (nscuro) [reply src/main/java/org/dependencytrack/tasks/metrics/PortfolioMetricsUpdateTask.java:89]: @sonatype-lift ignore

File: src/main/java/org/dependencytrack/persistence/StoredProcedures.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    /**
+     * Execute a given stored procedure and customize the execution.
+     *
+     * @param procedure     The {@link Procedure} to execute
+     * @param queryConsumer {@link Consumer} for customizing the {@link StoredProcedureQuery}
+     * @since 5.0.0
+     */
+    public static void execute(final Procedure procedure, final Consumer<StoredProcedureQuery> queryConsumer) {
+        try (final var qm = new QueryManager()) {
+            final Query<?> query = qm.getPersistenceManager().newQuery("STOREDPROC", procedure.quotedName());
```
[the remark this answers is not in the document]
Comment (nscuro) [reply src/main/java/org/dependencytrack/persistence/StoredProcedures.java:50]: Input is not user-controllable.

[the remark this answers is not in the document]
Comment (nscuro) [reply src/main/java/org/dependencytrack/persistence/StoredProcedures.java:50]: @sonatype-lift ignore

**Review by VithikaS — APPROVED** [review]
Looks really good! 
Looking forward to see how performance compares with the previous implementation.

**Pull request merged**